cloudca/resource_cloudca_vpn.go

```diff
@@ -0,0 +1,151 @@
+				ForceNew:    true,
+				Computed:    true,
+				Description: "ID of the public IP address associated with the vpn",
+			},
+			"state": {
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Computed:    true,
+				Description: "State of the vpn",
+			},
+			"type": {
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Computed:    true,
+				Description: "Type of vpn connection",
+			},
+		},
+	}
+}
+
+func resourceCloudcaVpnCreate(d *schema.ResourceData, meta interface{}) error {
+	vpnIPPurpose := "SOURCE_NAT"
+	ccaResources, rerr := getResourcesForEnvironmentID(meta.(*cca.CcaClient), d.Get("environment_id").(string))
+	if rerr != nil {
+		return rerr
+	}
+
+	var vpnPubIPID string
+	pubIps, _ := ccaResources.PublicIps.List()
+	for _, ip := range pubIps {
+		if ip.VpcId == d.Get("vpc_id").(string) {
+			for _, purpose := range ip.Purposes {
+				if purpose == vpnIPPurpose {
+					vpnPubIPID = ip.Id
+				}
+			}
+		}
+	}
+
+	if vpnPubIPID == "" {
+		return fmt.Errorf("Error enabling the VPN because no Source NAT IP was found for the VPC")
+	}
+
+	_, err := ccaResources.RemoteAccessVpn.Enable(vpnPubIPID)
+	if err != nil {
+		return fmt.Errorf("Error enabling the VPN: %s", err)
+	}
+	d.SetId(vpnPubIPID)
+	return resourceCloudcaVpnRead(d, meta)
+}
+
+func resourceCloudcaVpnRead(d *schema.ResourceData, meta interface{}) error {
+	ccaResources, rerr := getResourcesForEnvironmentID(meta.(*cca.CcaClient), d.Get("environment_id").(string))
+	if rerr != nil {
+		return rerr
+	}
+
+	vpn, err := ccaResources.RemoteAccessVpn.Get(d.Id())
+	if err != nil {
+		return handleNotFoundError("VPN", false, err, d)
+	}
+
+	if vpn.State == "Disabled" {
+		// If the VPN is disabled, it means the VPN is not active
+		// so this entity is "missing" (at least as far as terraform is concerned).
+		d.SetId("")
+		return nil
+	}
+	if err := d.Set("state", vpn.State); err != nil {
+		return fmt.Errorf("Error reading Trigger: %s", err)
+	}
+	if err := d.Set("certificate", vpn.Certificate); err != nil {
+		return fmt.Errorf("Error reading Trigger: %s", err)
+	}
+	if err := d.Set("preshared_key", vpn.PresharedKey); err != nil {
+		return fmt.Errorf("Error reading Trigger: %s", err)
+	}
+	if err := d.Set("public_ip", vpn.PublicIpAddress); err != nil {
+		return fmt.Errorf("Error reading Trigger: %s", err)
+	}
+	if err := d.Set("public_ip_id", vpn.PublicIpAddressId); err != nil {
+		return fmt.Errorf("Error reading Trigger: %s", err)
+	}
+	if err := d.Set("type", vpn.Type); err != nil {
+		return fmt.Errorf("Error reading Trigger: %s", err)
+	}
+	return nil
+}
+
+func resourceCloudcaVpnDelete(d *schema.ResourceData, meta interface{}) error {
+	ccaResources, rerr := getResourcesForEnvironmentID(meta.(*cca.CcaClient), d.Get("environment_id").(string))
+	if rerr != nil {
+		return rerr
+	}
+	if _, err := ccaResources.RemoteAccessVpn.Disable(d.Id()); err != nil {
+		return handleNotFoundError("VPN Delete", true, err, d)
```

Comment:
Is it necessary to error out if the delete returns a NotFound? Can we consider it successful instead?

Reply:
Good point.  I followed the example of how this was done in other areas of the plugin library.

If the disable errors though, we don't know the state.  Can we really assume it is now disabled and not throw an error?

Reply:
To be clear.  The delete is throwing an error (we are catching all errors).  We are returning a `not found` error, but that is not necessarily why if failed.

Reply:
We don't have to do it here, but it would be nice to process the error to see if it's a `not found` that way we could show it as successful

Reply:
catching the `404` errors and handling appropriately in that case now...  fixed...